File: apps/agent/tests/integration/update_key_add_remaining_test.go
package integration_test

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/require"
	"github.com/unkeyed/unkey/apps/agent/pkg/integration"
	"github.com/unkeyed/unkey/apps/agent/pkg/server"
)

func TestUpdateRemaining(t *testing.T) {
	createApiResponse := integration.Step[server.CreateApiResponse]{
		Name:   "Create API",
		Method: "POST",
		Url:    fmt.Sprintf("%s/v1/apis.createApi", BASE_URL),
		Header: map[string]string{
			"Content-Type":  "application/json",
			"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
		},
		Body: map[string]any{
			"name": "scenario-test-pls-delete",
		},
	}.Run(t)
	require.Equal(t, 200, createApiResponse.Status)
	require.NotEmpty(t, createApiResponse.Body.ApiId)
	require.NotEmpty(t, createApiResponse.Header.Get("Unkey-Trace-Id"))

	defer func() {

		deleteApiResponse := integration.Step[map[string]any]{
			Name:   "Delete API",
			Method: "POST",
			Url:    fmt.Sprintf("%s/v1/apis.deleteApi", BASE_URL),
			Header: map[string]string{
				"Content-Type":  "application/json",
				"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
			},
			Body: map[string]any{
				"apiId": createApiResponse.Body.ApiId,
			},
		}.Run(t)
		require.Equal(t, 200, deleteApiResponse.Status)
		require.NotEmpty(t, deleteApiResponse.Header.Get("Unkey-Trace-Id"))
	}()

	createKeyResponse := integration.Step[server.CreateKeyResponse]{
		Name:   "Create Key",
		Method: "POST",
		Url:    fmt.Sprintf("%s/v1/keys.createKey", BASE_URL),
		Header: map[string]string{
			"Content-Type":  "application/json",
			"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
		},
		Body: map[string]any{
			"apiId":     createApiResponse.Body.ApiId,
			"remaining": 5,
		},
	}.Run(t)
	require.Equal(t, 200, createKeyResponse.Status)
	require.NotEmpty(t, createKeyResponse.Body.Key)
	require.NotEmpty(t, createKeyResponse.Body.KeyId)
	require.NotEmpty(t, createKeyResponse.Header.Get("Unkey-Trace-Id"))

	defer func() {
		deleteKeyResponse := integration.Step[map[string]any]{
			Name:   "Revoke Key",
			Method: "POST",
			Url:    fmt.Sprintf("%s/v1/keys.deleteKey", BASE_URL),
			Header: map[string]string{
				"Content-Type":  "application/json",
				"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
			},
			Body: map[string]any{
				"keyId": createKeyResponse.Body.KeyId,
			},
		}.Run(t)
		require.Equal(t, 200, deleteKeyResponse.Status)
		require.NotEmpty(t, deleteKeyResponse.Header.Get("Unkey-Trace-Id"))

	}()
	// Use up all 5 verifications
	for i := 4; i >= 0; i-- {

		verifyKeyRespone := integration.Step[server.VerifyKeyResponseV1]{
			Name:   "Verify Key",
			Method: "POST",
			Url:    fmt.Sprintf("%s/v1/keys.verifyKey", BASE_URL),
			Header: map[string]string{
				"Content-Type": "application/json",
			},
			Body: map[string]any{
				"key": createKeyResponse.Body.Key,
			},
		}.Run(t)
		require.Equal(t, 200, verifyKeyRespone.Status)
		require.True(t, verifyKeyRespone.Body.Valid)
		require.Equal(t, int32(i), *verifyKeyRespone.Body.Remaining)
		require.NotEmpty(t, verifyKeyRespone.Header.Get("Unkey-Trace-Id"))
	}
	failingVerifyKeyResponse := integration.Step[server.VerifyKeyResponseV1]{
		Name:   "Verify Key - should fail",
		Method: "POST",
		Url:    fmt.Sprintf("%s/v1/keys.verifyKey", BASE_URL),
		Header: map[string]string{
			"Content-Type": "application/json",
		},
		Body: map[string]any{
			"key": createKeyResponse.Body.Key,
		},
	}.Run(t)
	require.Equal(t, 200, failingVerifyKeyResponse.Status)
	require.False(t, failingVerifyKeyResponse.Body.Valid)
	require.Equal(t, int32(0), *failingVerifyKeyResponse.Body.Remaining)
	require.Equal(t, "KEY_USAGE_EXCEEDED", failingVerifyKeyResponse.Body.Code)
	require.NotEmpty(t, failingVerifyKeyResponse.Header.Get("Unkey-Trace-Id"))

	// Add 5 more verifications
	updateKeyResponse := integration.Step[server.UpdateKeyResponse]{
		Name:   "Set remaining to 5",
		Method: "POST",
		Url:    fmt.Sprintf("%s/v1/keys.updateKey", BASE_URL),
		Header: map[string]string{
			"Content-Type":  "application/json",
			"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
		},
		Body: map[string]any{
			"keyId":     createKeyResponse.Body.KeyId,
			"remaining": 5,
		},
	}.Run(t)
	require.Equal(t, 200, updateKeyResponse.Status)
	require.NotEmpty(t, updateKeyResponse.Header.Get("Unkey-Trace-Id"))

	// Verify the key has new remaining
	verifyKeyResponse := integration.Step[server.VerifyKeyResponseV1]{
		Name:   "Verify key after update",
		Method: "POST",
		Url:    fmt.Sprintf("%s/v1/keys.verifyKey", BASE_URL),
		Header: map[string]string{
			"Content-Type": "application/json",
		},
		Body: map[string]any{
			"key": createKeyResponse.Body.Key,
		},
	}.Run(t)
	require.Equal(t, 200, verifyKeyResponse.Status)
	require.True(t, verifyKeyResponse.Body.Valid)
	require.Equal(t, int32(4), *verifyKeyResponse.Body.Remaining)
	require.NotEmpty(t, verifyKeyResponse.Header.Get("Unkey-Trace-Id"))

}

File: web/apps/api/src/integration/update_key_add_remaining.test.ts
import { IntegrationHarness } from "@/pkg/testutil/integration-harness";
import type { V1ApisCreateApiRequest, V1ApisCreateApiResponse } from "@/routes/v1_apis_createApi";
import type { V1ApisDeleteApiRequest, V1ApisDeleteApiResponse } from "@/routes/v1_apis_deleteApi";
import type { V1KeysCreateKeyRequest, V1KeysCreateKeyResponse } from "@/routes/v1_keys_createKey";
import type {
  V1KeysUpdateRemainingRequest,
  V1KeysUpdateRemainingResponse,
} from "@/routes/v1_keys_updateRemaining";
import type { V1KeysVerifyKeyRequest, V1KeysVerifyKeyResponse } from "@/routes/v1_keys_verifyKey";
import { describe, expect, test } from "vitest";

describe("some", () => {
  test("update a key's remaining limit", async (t) => {
    const h = await IntegrationHarness.init(t);
    const { key: rootKey } = await h.createRootKey(["*"]);

    const createApiResponse = await h.post<V1ApisCreateApiRequest, V1ApisCreateApiResponse>({
      url: `${h.baseUrl}/v1/apis.createApi`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        name: "scenario-test-pls-delete",
      },
    });
    expect(createApiResponse.status).toEqual(200);
    expect(createApiResponse.body.apiId).toBeDefined();
    expect(createApiResponse.headers).toHaveProperty("unkey-request-id");

    const createKeyResponse = await h.post<V1KeysCreateKeyRequest, V1KeysCreateKeyResponse>({
      url: `${h.baseUrl}/v1/keys.createKey`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        apiId: createApiResponse.body.apiId,
        byteLength: 32,
        prefix: "test",
        remaining: 5,
        enabled: true,
      },
    });
    expect(createKeyResponse.status).toEqual(200);

    for (let i = 4; i >= 0; i--) {
      const valid = await h.post<V1KeysVerifyKeyRequest, V1KeysVerifyKeyResponse>({
        url: `${h.baseUrl}/v1/keys.verifyKey`,
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${rootKey}`,
        },
        body: {
          apiId: createApiResponse.body.apiId,
          key: createKeyResponse.body.key,
        },
      });

      expect(valid.status).toEqual(200);
      expect(valid.body.valid).toBe(true);
      expect(valid.body.remaining).toEqual(i);
    }

    const invalid = await h.post<V1KeysVerifyKeyRequest, V1KeysVerifyKeyResponse>({
      url: `${h.baseUrl}/v1/keys.verifyKey`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        apiId: createApiResponse.body.apiId,
        key: createKeyResponse.body.key,
      },
    });
    expect(invalid.status).toEqual(200);
    expect(invalid.body.valid).toBe(false);
    expect(invalid.body.remaining).toEqual(0);

    const updateKeyResponse = await h.post<
      V1KeysUpdateRemainingRequest,
      V1KeysUpdateRemainingResponse
    >({
      url: `${h.baseUrl}/v1/keys.updateRemaining`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        keyId: createKeyResponse.body.keyId,
        op: "increment",
        value: 5,
      },
    });

    expect(updateKeyResponse.status).toEqual(200);
    expect(updateKeyResponse.body.remaining).toEqual(5);

    const validAfterUpdate = await h.post<V1KeysVerifyKeyRequest, V1KeysVerifyKeyResponse>({
      url: `${h.baseUrl}/v1/keys.verifyKey`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        apiId: createApiResponse.body.apiId,
        key: createKeyResponse.body.key,
      },
    });
    expect(validAfterUpdate.status).toEqual(200);
    expect(validAfterUpdate.body.valid).toBe(true);
    expect(validAfterUpdate.body.remaining).toEqual(4);

    /**
     * Teardown
     */
    const deleteApi = await h.post<V1ApisDeleteApiRequest, V1ApisDeleteApiResponse>({
      url: `${h.baseUrl}/v1/apis.deleteApi`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        apiId: createApiResponse.body.apiId,
      },
    });
    expect(deleteApi.status).toEqual(200);
  });
});
